optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/DroolsBiCondition.java

```diff
@@ -62,31 +74,21 @@ public DroolsBiCondition(DroolsMetadata<Object, A> aMetadata, DroolsMetadata<Obj
     }
 
     public DroolsBiCondition<A, B> andFilter(BiPredicate<A, B> predicate) {
-        PatternDSL.PatternDef<Object> newPattern = bMetadata.getPattern()
+        Supplier<PatternDSL.PatternDef<Object>> patternSupplier = () -> bMetadata.buildPattern()
                 .expr(aMetadata.getVariableDeclaration(),
                         (b, a) -> predicate.test(aMetadata.extract(a), bMetadata.extract(b)));
-        if (bMetadata instanceof DroolsInferredMetadata) {
-            return new DroolsBiCondition<>(aMetadata, ((DroolsInferredMetadata) bMetadata).substitute(newPattern));
-        } else {
-            return new DroolsBiCondition<>(aMetadata, ((DroolsGenuineMetadata) bMetadata).substitute(newPattern));
-        }
+        return new DroolsBiCondition<>(aMetadata, bMetadata.substitute(patternSupplier));
     }
 
     public <C> DroolsTriCondition<A, B, C> andJoin(DroolsUniCondition<C> cCondition,
             AbstractTriJoiner<A, B, C> triJoiner) {
         DroolsMetadata<Object, C> cMetadata = cCondition.getAMetadata();
-        PatternDSL.PatternDef<Object> newPattern = cMetadata.getPattern()
+        Supplier<PatternDSL.PatternDef<Object>> newPattern = () -> cMetadata.buildPattern()
                 .expr(UUID.randomUUID().toString(), aMetadata.getVariableDeclaration(),
```

Comment:
There are still usages of UUID - Is that intended?

Reply:
The Drools API requires a textual ID for the expression. As far as I know, it doesn't make it into the rule and may be completely random. But I'll look around to see if I can somehow get rid of it.